core/src/main/scala/org/apache/spark/storage/DiskBlockManager.scala

```diff
@@ -53,35 +53,98 @@ private[spark] class DiskBlockManager(blockManager: BlockManager, conf: SparkCon
 
   private val shutdownHook = addShutdownHook()
 
+  private abstract class FileAllocationStrategy {
+    def apply(filename: String): File
+
+    protected def getFile(filename: String, storageDirs: Array[File]): File = {
+      require(storageDirs.nonEmpty, "could not find file when the directories are empty")
+
+      // Figure out which local directory it hashes to, and which subdirectory in that
+      val hash = Utils.nonNegativeHash(filename)
+      val dirId = localDirs.indexOf(storageDirs(hash % storageDirs.length))
+      val subDirId = (hash / storageDirs.length) % subDirsPerLocalDir
+
+      // Create the subdirectory if it doesn't already exist
+      val subDir = subDirs(dirId).synchronized {
+        val old = subDirs(dirId)(subDirId)
+        if (old != null) {
+          old
+        } else {
+          val newDir = new File(localDirs(dirId), "%02x".format(subDirId))
+          if (!newDir.exists() && !newDir.mkdir()) {
+            throw new IOException(s"Failed to create local dir in $newDir.")
+          }
+          subDirs(dirId)(subDirId) = newDir
+          newDir
+        }
+      }
+
+      new File(subDir, filename)
+    }
+  }
+
   /** Looks up a file by hashing it into one of our local subdirectories. */
   // This method should be kept in sync with
   // org.apache.spark.network.shuffle.ExternalShuffleBlockResolver#getFile().
-  def getFile(filename: String): File = {
-    // Figure out which local directory it hashes to, and which subdirectory in that
-    val hash = Utils.nonNegativeHash(filename)
-    val dirId = hash % localDirs.length
-    val subDirId = (hash / localDirs.length) % subDirsPerLocalDir
-
-    // Create the subdirectory if it doesn't already exist
-    val subDir = subDirs(dirId).synchronized {
-      val old = subDirs(dirId)(subDirId)
-      if (old != null) {
-        old
-      } else {
-        val newDir = new File(localDirs(dirId), "%02x".format(subDirId))
-        if (!newDir.exists() && !newDir.mkdir()) {
-          throw new IOException(s"Failed to create local dir in $newDir.")
-        }
-        subDirs(dirId)(subDirId) = newDir
-        newDir
+  private object hashAllocator extends FileAllocationStrategy {
+    def apply(filename: String): File = getFile(filename, localDirs)
+  }
+
+  /** Looks up a file by hierarchy way in different speed storage devices. */
+  private val hierarchyStore = conf.getOption("spark.storage.hierarchyStore")
+  private class HierarchyAllocator extends FileAllocationStrategy {
+    case class LayerInfo(key: String, threshold: Long, dirs: Array[File])
+    val hsSpecs: Array[(String, Long)] =
+      // e.g.: hierarchyStore = "ssd 200GB, hdd 100GB"
+      hierarchyStore.get.trim.split(",").map {
+        s => val x = s.trim.split(" +")
+          (x(0).toLowerCase, Utils.byteStringAsGb(x(1)))
```

Comment:
Ok, by this way, we can even eliminate "file.getParentFile.getUsableSpace>>30" operation in apply(), thanks!